https://github.com/AnetaPopelova/czechitas-python-pro-datovou-analyzu-2024-09/blob/main/notebooks/lekce-06/alcohol-zadani.md

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
data_alkohol = pd.read_csv("Stats_survey.csv")
data_alkohol


Timestamp Your Sex?  \
0     2024/03/07 5:12:01 pm EET    Female   
1     2024/03/07 5:12:08 pm EET      Male   
2     2024/03/07 5:12:25 pm EET      Male   
3     2024/03/07 5:12:28 pm EET      Male   
4     2024/03/07 5:13:00 pm EET    Female   
..                          ...       ...   
401  2024/03/12 11:05:33 am EET    Female   
402   2024/03/12 3:26:14 pm EET    Female   
403  2024/03/13 10:32:19 pm EET    Female   
404  2024/03/13 10:32:27 pm EET      Male   
405  2024/03/16 10:30:36 pm EET    Female   

     Your Matric (grade 12) Average/ GPA (in %)  \
0                                          76.0   
1                                          89.0   
2                                          76.0   
3                                          89.0   
4                                          74.0   
..                                          ...   
401                                        74.0   
402                                        73.0   
403                                        86.0   
404                                        85.0   
405                                        80.0   

    What year were you in last year (2023) ?  \
0                                   2nd Year   
1                                   2nd Year   
2                                   1st Year   
3                                   2nd Year   
4                                   2nd Year   
..                                       ...   
401                                      NaN   
402                                 2nd Year   
403                                 1st Year   
404                                      NaN   
405                                 1st Year   

    What faculty does your degree fall under?  \
0                      Arts & Social Sciences   
1              Economic & Management Sciences   
2                                AgriSciences   
3                                 Engineering   
4                      Arts & Social Sciences   
..                                        ...   
401                                   Science   
402            Economic & Management Sciences   
403                                       NaN   
404            Economic & Management Sciences   
405                                       NaN   

     Your 2023 academic year average/GPA in % (Ignore if you are 2024 1st year student)  \
0                                                 72.0                                    
1                                                 75.0                                    
2                                                 55.0                                    
3                                                 84.0                                    
4                                                 52.0                                    
..                                                 ...                                    
401                                                NaN                                    
402                                               58.0                                    
403                                                NaN                                    
404                                                NaN                                    
405                                               63.0                                    

          Your Accommodation Status Last Year (2023)  \
0    Private accommodation/ stay with family/friends   
1    Private accommodation/ stay with family/friends   
2    Private accommodation/ stay with family/friends   
3    Private accommodation/ stay with family/friends   
4    Private accommodation/ stay with family/friends   
..                                               ...   
401  Private accommodation/ stay with family/friends   
402  Private accommodation/ stay with family/friends   
403  Private accommodation/ stay with family/friends   
404  Private accommodation/ stay with family/frie

In [4]:
data_alkohol.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 17 columns):
 #   Column                                                                                             Non-Null Count  Dtype  
---  ------                                                                                             --------------  -----  
 0   Timestamp                                                                                          406 non-null    object 
 1   Your Sex?                                                                                          404 non-null    object 
 2   Your Matric (grade 12) Average/ GPA (in %)                                                         399 non-null    float64
 3   What year were you in last year (2023) ?                                                           333 non-null    object 
 4   What faculty does your degree fall under?                                                          399 non-null    object 

In [5]:
data_alkohol.rename(
    columns={
      "Timestamp": "Timestamp",
        "Your Sex?": "Sex",
        "Your Matric (grade 12) Average/ GPA (in %)": "Matric_GPA",
        "What year were you in last year (2023) ?": "Year",
        "What faculty does your degree fall under?": "Faculty",
        "Your 2023 academic year average/GPA in % (Ignore if you are 2024 1st year student)": "Yearly_GPA",
        "Your Accommodation Status Last Year (2023)": "Accommodation",
        "Monthly Allowance in 2023": "Monthly_Allowance",
        "Were you on scholarship/bursary in 2023?": "Scholarship",
        "Additional amount of studying (in hrs) per week": "Study_Hours",
        "How often do you go out partying/socialising during the week? ": "Socialising_Frequency",
        "On a night out, how many alcoholic drinks do you consume?": "Drinks_Per_Night",
        "How many classes do you miss per week due to alcohol reasons, (i.e: being hungover or too tired?)": "Missed_Classes",
        "How many modules have you failed thus far into your studies?": "Modules_Failed",
        "Are you currently in a romantic relationship?": "In_Relationship",
        "Do your parents approve alcohol consumption?": "Parents_Approve",
        "How strong is your relationship with your parent/s?": "Relationship_Parents",
    }, 
    inplace=True,
)
data_alkohol

Timestamp     Sex  Matric_GPA      Year  \
0     2024/03/07 5:12:01 pm EET  Female        76.0  2nd Year   
1     2024/03/07 5:12:08 pm EET    Male        89.0  2nd Year   
2     2024/03/07 5:12:25 pm EET    Male        76.0  1st Year   
3     2024/03/07 5:12:28 pm EET    Male        89.0  2nd Year   
4     2024/03/07 5:13:00 pm EET  Female        74.0  2nd Year   
..                          ...     ...         ...       ...   
401  2024/03/12 11:05:33 am EET  Female        74.0       NaN   
402   2024/03/12 3:26:14 pm EET  Female        73.0  2nd Year   
403  2024/03/13 10:32:19 pm EET  Female        86.0  1st Year   
404  2024/03/13 10:32:27 pm EET    Male        85.0       NaN   
405  2024/03/16 10:30:36 pm EET  Female        80.0  1st Year   

                            Faculty  Yearly_GPA  \
0            Arts & Social Sciences        72.0   
1    Economic & Management Sciences        75.0   
2                      AgriSciences        55.0   
3                       Engineering        84.0   
4            Arts & Social Sciences        52.0   
..                              ...         ...   
401                         Science         NaN   
402  Economic & Management Sciences        58.0   
403                             NaN         NaN   
404  Economic & Management Sciences         NaN   
405                             NaN        63.0   

                                       Accommodation Monthly_Allowance  \
0    Private accommodation/ stay with family/friends    R 4001- R 5000   
1    Private accommodation/ stay with family/friends   R 7001 - R 8000   
2    Private accommodation/ stay with family/friends    R 4001- R 5000   
3    Private accommodation/ stay with family/friends   R 6001 - R 7000   
4    Private accommodation/ stay with family/friends    R 4001- R 5000   
..                                               ...               ...   
401  Private accommodation/ stay with family/friends               NaN   
402  Private accommodation/ stay with family/friends   R 6001 - R 7000   
403  Private accommodation/ stay with family/friends    R 4001- R 5000   
404  Private accommodation/ stay with family/friends    R 4001- R 5000   
405  Private accommodation/ stay with family/friends   R 5001 - R 6000   

             Scholarship Study_Hours Socialising_Frequency Drinks_Per_Night  \
0                     No          8+         Only weekends               8+   
1    Yes (NSFAS, etc...)          8+         Only weekends              3-5   
2                     No         3-5                     2               8+   
3                     No         3-5                     3               8+   
4                     No         3-5         Only weekends              5-8   
..                   ...         ...                   ...              ...   
401                   No         3-5                     3              1-3   
402  Yes (NSFAS, etc...)         5-8                     1               8+   
403                   No         5-8                     2              3-5   
404                   No         1-3                    4+              5-8   
405                   No         3-5         Only weekends              3-5   

    Missed_Classes Modules_Failed In_Relationship Parents_Approve  \
0                3              0             Yes             Yes   
1               4+              0              No             Yes   
2                3              0              No             Yes   
3                2              0             Yes             Yes   
4                1              3              No             Yes   
..             ...            ...             ...             ...   
401              0              0              No             Yes   
402              0              0             Yes             Yes   
403              1              0              No             Yes   
404             4+              0              No             Yes   
405              1             

In [6]:
data_alkohol.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Timestamp              406 non-null    object 
 1   Sex                    404 non-null    object 
 2   Matric_GPA             399 non-null    float64
 3   Year                   333 non-null    object 
 4   Faculty                399 non-null    object 
 5   Yearly_GPA             320 non-null    float64
 6   Accommodation          383 non-null    object 
 7   Monthly_Allowance      375 non-null    object 
 8   Scholarship            398 non-null    object 
 9   Study_Hours            403 non-null    object 
 10  Socialising_Frequency  404 non-null    object 
 11  Drinks_Per_Night       404 non-null    object 
 12  Missed_Classes         403 non-null    object 
 13  Modules_Failed         403 non-null    object 
 14  In_Relationship        403 non-null    object 
 15  Parent

Počet dotazovaných 406
Ze všech sloupečků nejvícekrát neodpověděli tázaní, ve kterém ročníku byli minulý rok (V roce 2023) - 333
Dále nejméně dotazovaných odpovědělo, jaký mělo za rok 2023 průměr - 320

In [9]:
print("\nPočet záznamů v každém sloupci:")
print(data_alkohol.count())


Počet záznamů v každém sloupci:
Timestamp                406
Sex                      404
Matric_GPA               399
Year                     333
Faculty                  399
Yearly_GPA               320
Accommodation            383
Monthly_Allowance        375
Scholarship              398
Study_Hours              403
Socialising_Frequency    404
Drinks_Per_Night         404
Missed_Classes           403
Modules_Failed           403
In_Relationship          403
Parents_Approve          402
Relationship_Parents     403
dtype: int64


In [14]:
print("Datové typy jednotlivých sloupců")
for column in data_alkohol.columns:
    print(f"{column}: {data_alkohol[column].dtype}")



Datové typy jednotlivých sloupců
Timestamp: object
Sex: object
Matric_GPA: float64
Year: object
Faculty: object
Yearly_GPA: float64
Accommodation: object
Monthly_Allowance: object
Scholarship: object
Study_Hours: object
Socialising_Frequency: object
Drinks_Per_Night: object
Missed_Classes: object
Modules_Failed: object
In_Relationship: object
Parents_Approve: object
Relationship_Parents: object


In [12]:
print("\nPočet záznamů v každém sloupci:")
for column in data_alkohol.columns:
    print(f"{column}: {data_alkohol[column].count()}")


Počet záznamů v každém sloupci:
Timestamp: 406
Sex: 404
Matric_GPA: 399
Year: 333
Faculty: 399
Yearly_GPA: 320
Accommodation: 383
Monthly_Allowance: 375
Scholarship: 398
Study_Hours: 403
Socialising_Frequency: 404
Drinks_Per_Night: 404
Missed_Classes: 403
Modules_Failed: 403
In_Relationship: 403
Parents_Approve: 402
Relationship_Parents: 403


In [25]:
print("\nPočet unikátních záznamů'):")
for column in data_alkohol.columns:
    if column != 'Timestamp':
        total_count = data_alkohol[column].count()
        unique_count = data_alkohol[column].nunique
        duplicates_removed = total_count - unique_count
        print(f"{column}: {duplicates_removed}")


Počet unikátních záznamů'):
Sex: 402
Matric_GPA: 356
Year: 328
Faculty: 391
Yearly_GPA: 275
Accommodation: 381
Monthly_Allowance: 370
Scholarship: 396
Study_Hours: 398
Socialising_Frequency: 398
Drinks_Per_Night: 399
Missed_Classes: 398
Modules_Failed: 398
In_Relationship: 401
Parents_Approve: 400
Relationship_Parents: 399


In [13]:
print("\nPočet chybějících hodnot v každém sloupci:")
for column in data_alkohol.columns:
    missing_count = data_alkohol[column].isnull().sum()
    print(f"{column}: {missing_count}")


Počet chybějících hodnot v každém sloupci:
Timestamp: 0
Sex: 2
Matric_GPA: 7
Year: 73
Faculty: 7
Yearly_GPA: 86
Accommodation: 23
Monthly_Allowance: 31
Scholarship: 8
Study_Hours: 3
Socialising_Frequency: 2
Drinks_Per_Night: 2
Missed_Classes: 3
Modules_Failed: 3
In_Relationship: 3
Parents_Approve: 4
Relationship_Parents: 3


In [27]:
#ženy - faktula - průměr známek
data_women = data_alkohol[data_alkohol['Sex']== 'Female']
faculty_counts = data_women['Faculty'].value_counts().mean()
print("Průměrný počet žen na každé fakultě:")
print(faculty_counts)

Průměrný počet žen na každé fakultě:
22.875


In [37]:
faculty_counts = data_women['Faculty'].value_counts()
print("\nPočet žena na každé faktultě")
print(faculty_counts)


Počet žena na každé faktultě
Faculty
Economic & Management Sciences    85
Arts & Social Sciences            31
Science                           27
Engineering                       12
AgriSciences                       8
Medicine and Health Services       8
Law                                6
Education                          6
Name: count, dtype: int64


In [38]:
average_gpa_by_faculty = data_women.groupby('Faculty')['Yearly_GPA'].mean()
print("\nPrůměrné známky Yearly_GPA pro každou fakultu:")
print(average_gpa_by_faculty)
worst_faculty = average_gpa_by_faculty.idxmin()
worst_gpa = average_gpa_by_faculty.min()
print(f"\nFaktula s nejhorším průměrným GPA: {worst_faculty} s průměrem {worst_gpa:.2f}")



Průměrné známky Yearly_GPA pro každou fakultu:
Faculty
AgriSciences                      64.833333
Arts & Social Sciences            67.750000
Economic & Management Sciences    67.390417
Education                         71.200000
Engineering                       62.825000
Law                               65.200000
Medicine and Health Services      73.385714
Science                           69.750000
Name: Yearly_GPA, dtype: float64

Faktula s nejhorším průměrným GPA: Engineering s průměrem 62.83


Nejhorší průměr má mezi ženami Engineering = Inženýrství, nejlepší průměr Medicine and Health Services a Education = Medicína a zdravotnictví a Vzdělávání. V obou odvětvích ale studuje jen několik žen. Nejvíce žen studuje Ekonomiku a management, stejně jako muži. Ženy mají v Ekonmice a Umění velmi podobný průměr známek.

In [34]:
# ženy a nejčastější hodnota studijního času
Female_Study_Hours = data_women['Study_Hours'].mode()[0]
print("\nNejčastější hodnota v čase věnovaném studiu pro ženy:")
print(Female_Study_Hours)


Nejčastější hodnota v čase věnovaném studiu pro ženy:
8+


Ženy opravdu hodně věnují studiu více než 8 hodin týdně.

In [28]:
average_gpa = data_women['Yearly_GPA'].mean()
print("\nPrůměrný roční GPA žen:")
print(average_gpa)


Průměrný roční GPA žen:
67.59606666666667


In [36]:
def convert_socialising_frequency(freq_str):
    if freq_str == 'only weekends':
        return 2  # Představuje 2 dny
    elif freq_str == '4+':
        return 5  # Představuje 4 a více 
    else:
        try:
            return int(freq_str)  
        except ValueError:
            return None 

data_women['Socialising_Frequency'] = data_women['Socialising_Frequency'].apply(convert_socialising_frequency)

average_socialising_frequency = data_women['Socialising_Frequency'].mean()
print("\nPrůměrná frekvence socializace pro ženy:")
print(average_socialising_frequency)


Průměrná frekvence socializace pro ženy:
1.859504132231405


C:\Users\zvitk\AppData\Local\Temp\ipykernel_4104\2359474136.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_women['Socialising_Frequency'] = data_women['Socialising_Frequency'].apply(convert_socialising_frequency)


In [41]:
data_men = data_alkohol[data_alkohol['Sex'] == 'Male']
men_counts_and_gpa = data_men.groupby('Faculty').agg(
    male_count=('Sex', 'count'),
    average_gpa=('Yearly_GPA', 'mean')
    ).reset_index()

print("\nPočet mužů na každé fakultě a jejich průměrné Yearly_GPA:")
print(men_counts_and_gpa)


Počet mužů na každé fakultě a jejich průměrné Yearly_GPA:
                          Faculty  male_count  average_gpa
0                    AgriSciences          14    64.307692
1          Arts & Social Sciences          16    63.363636
2  Economic & Management Sciences         126    65.415000
3                     Engineering          25    66.176471
4                             Law           4    64.250000
5    Medicine and Health Services           2    61.500000
6                         Science          29    64.652174


Muži mají ve všech zaměřeních fakult podobný průměr známek, nejvíce mužů studuje Ekonomiku a management. Je to opravdu velký počet mužů oproti jiným fakultám. V medicíně a zdravotnictví studují jen 2 muži, což je menší počet než u žen v tomto oboru studiu.

In [43]:
def convert_socialising_frequency(freq_str):
    if freq_str == 'only weekends':
        return 2  # Představuje 2 dny
    elif freq_str == '4+':
        return 5  # Představuje 4 a více
    else:
        try:
            return int(freq_str) 
        except ValueError:
            return None  

data_men = data_alkohol[data_alkohol['Sex'] == 'Male']

data_men['Socialising_Frequency'] = data_men['Socialising_Frequency'].apply(convert_socialising_frequency)

average_socialising_frequency_men = data_men['Socialising_Frequency'].mean()
print("\nPrůměrná frekvence socializace pro muže:")
print(average_socialising_frequency_men)


Průměrná frekvence socializace pro muže:
1.9176470588235295


C:\Users\zvitk\AppData\Local\Temp\ipykernel_4104\136433025.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_men['Socialising_Frequency'] = data_men['Socialising_Frequency'].apply(convert_socialising_frequency)


In [45]:
def convert_study_hours(hours_str):
    if hours_str == '1-3':
        return 2  # Průměr mezi 1 a 3
    elif hours_str == '3-5':
        return 4  # Průměr mezi 3 a 5
    elif hours_str == '5-8':
        return 6.5  # Průměr mezi 5 a 8
    elif hours_str == '8+':
        return 8 
    else:
        return None  

data_men = data_alkohol[data_alkohol['Sex'] == 'Male']

data_men['Study_Hours'] = data_men['Study_Hours'].apply(convert_study_hours)

most_common_study_hours = data_men['Study_Hours'].mode()[0]
print(f"\nNejčastější hodnota studijních hodin pro muže: {most_common_study_hours}")


Nejčastější hodnota studijních hodin pro muže: 8.0


C:\Users\zvitk\AppData\Local\Temp\ipykernel_4104\773036899.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_men['Study_Hours'] = data_men['Study_Hours'].apply(convert_study_hours)


I Muži věnují studiu 8 a více hodin za týden. Což je stejné jako u žen

In [47]:
def convert_drinks(drinks_str):
    if drinks_str == '1-3':
        return 2  # Průměr mezi 1 a 3
    elif drinks_str == '3-5':
        return 4  # Průměr mezi 3 a 5
    elif drinks_str == '5-8':
        return 6.5  # Průměr mezi 5 a 8
    elif drinks_str == '8+':
        return 8 
    else:
        return None 

data_women = data_alkohol[data_alkohol['Sex'] == 'Female']
data_men = data_alkohol[data_alkohol['Sex'] == 'Male']

data_women['Drinks_Per_Night'] = data_women['Drinks_Per_Night'].apply(convert_drinks)
data_men['Drinks_Per_Night'] = data_men['Drinks_Per_Night'].apply(convert_drinks)

average_drinks_women = data_women['Drinks_Per_Night'].mean()
average_drinks_men = data_men['Drinks_Per_Night'].mean()

print(f"\nPrůměrný počet drinků za večer pro ženy: {average_drinks_women:.2f}")
print(f"Průměrný počet drinků za večer pro muže: {average_drinks_men:.2f}")


Průměrný počet drinků za večer pro ženy: 4.13
Průměrný počet drinků za večer pro muže: 6.07


C:\Users\zvitk\AppData\Local\Temp\ipykernel_4104\2508505739.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_women['Drinks_Per_Night'] = data_women['Drinks_Per_Night'].apply(convert_drinks)
C:\Users\zvitk\AppData\Local\Temp\ipykernel_4104\2508505739.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_men['Drinks_Per_Night'] = data_men['Drinks_Per_Night'].apply(convert_drinks)


Jak se dalo očekávat, ženy za večer vypijí méně drinků než muži.
Muži i ženy chodí do společnosti průměrně 1-2x týdně, spíše 2x.


zkusim poslat na GITHUB. 1.pokus se nezdařil. zkusím to znovu